pkg/build/controller/buildpod/factory.go

```diff
@@ -0,0 +1,213 @@
+package buildpod
+
+import (
+	"errors"
+	"time"
+
+	"github.com/golang/glog"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kcontroller "k8s.io/kubernetes/pkg/controller"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildclient "github.com/openshift/origin/pkg/build/client"
+	"github.com/openshift/origin/pkg/build/controller/policy"
+	strategy "github.com/openshift/origin/pkg/build/controller/strategy"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
+)
+
+const (
+	// We must avoid processing build pods until the build and pod stores have synced.
+	// If they haven't synced, to avoid a hot loop, we'll wait this long between checks.
+	storeSyncedPollPeriod = 100 * time.Millisecond
+	maxRetries            = 60
```

Comment:
> how do i know something is missing from it?

You have a build A that depends on a pod A-pod. The build controller is going down, A-pod is getting deleted for some reason. The build controller comes up, it reconciles all builds, sees that for A -> A-pod does not exist. This also depends on the state of the Build, for example, you know it was deleted because the Build is Running. You probably want to recreate the Pod if the Build is New or Pending.

> why? won't we still get a real delete event when the item is actually deleted from etcd? and of course if we had reconcile logic, we'd catch it then too.

I think you simply don't want to handle that event as ADD, not sure about the timing the DELETE is propagated from etcd (marked as deleted vs really deleted). You also need the logic to exist in the reconcile loop eg. for avoiding actions like adoption.